File: backend/src/Entity/StatusHistory.php
<?php

namespace App\Entity;

use App\Enum\ApplicationStatus;
use App\Repository\StatusHistoryRepository;
use Doctrine\DBAL\Types\Types;
use Doctrine\ORM\Mapping as ORM;
use Symfony\Component\Serializer\Attribute\Groups;
use Symfony\Component\Uid\Uuid;

#[ORM\Entity(repositoryClass: StatusHistoryRepository::class)]
class StatusHistory
{
    #[ORM\Id]
    #[ORM\Column(type: 'uuid', unique: true)]
    #[ORM\GeneratedValue(strategy: 'CUSTOM')]
    #[ORM\CustomIdGenerator(class: 'doctrine.uuid_generator')]
    #[Groups(['application:read'])]
    private ?Uuid $id = null;

    #[ORM\ManyToOne(targetEntity: Application::class, inversedBy: 'statusHistories')]
    #[ORM\JoinColumn(nullable: false)]
    private ?Application $application = null;

    #[ORM\Column(enumType: ApplicationStatus::class, nullable: true)]
    #[Groups(['application:read'])]
    private ?ApplicationStatus $previousStatus = null;

    #[ORM\Column(enumType: ApplicationStatus::class)]
    #[Groups(['application:read'])]
    private ?ApplicationStatus $newStatus = null;

    #[ORM\Column]
    #[Groups(['application:read'])]
    private ?\DateTimeImmutable $changedAt = null;

    #[ORM\Column(length: 20)]
    #[Groups(['application:read'])]
    private string $trigger = 'manual'; // 'manual' | 'auto_mail'

    #[ORM\Column(type: Types::TEXT, nullable: true)]
    #[Groups(['application:read'])]
    private ?string $notes = null;

    public function __construct()
    {
        $this->changedAt = new \DateTimeImmutable();
    }

    public function getId(): ?Uuid { return $this->id; }

    public function getApplication(): ?Application { return $this->application; }
    public function setApplication(?Application $v): static { $this->application = $v; return $this; }

    public function getPreviousStatus(): ?ApplicationStatus { return $this->previousStatus; }
    public function setPreviousStatus(?ApplicationStatus $v): static { $this->previousStatus = $v; return $this; }

    public function getNewStatus(): ?ApplicationStatus { return $this->newStatus; }
    public function setNewStatus(ApplicationStatus $v): static { $this->newStatus = $v; return $this; }

    public function getChangedAt(): ?\DateTimeImmutable { return $this->changedAt; }

    public function getTrigger(): string { return $this->trigger; }
    public function setTrigger(string $v): static { $this->trigger = $v; return $this; }

    public function getNotes(): ?string { return $this->notes; }
    public function setNotes(?string $v): static { $this->notes = $v; return $this; }
}

File: backend-java/src/main/java/com/jobdesk/domain/StatusHistory.java
package com.jobdesk.domain;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.PrePersist;
import jakarta.persistence.Table;

import java.time.LocalDateTime;
import java.util.UUID;

@Entity
@Table(name = "status_history")
public class StatusHistory {

    @Id
    private UUID id;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "application_id", nullable = false)
    private Application application;

    @Enumerated(EnumType.STRING)
    private ApplicationStatus previousStatus;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private ApplicationStatus newStatus;

    @Column(nullable = false)
    private LocalDateTime changedAt;

    @Column(length = 20, nullable = false)
    private String trigger = "manual";

    @Column(columnDefinition = "text")
    private String notes;

    @PrePersist
    void onPrePersist() {
        if (id == null) {
            id = UUID.randomUUID();
        }
        if (changedAt == null) {
            changedAt = LocalDateTime.now();
        }
    }

    public UUID getId() {
        return id;
    }

    public Application getApplication() {
        return application;
    }

    public void setApplication(Application application) {
        this.application = application;
    }

    public ApplicationStatus getPreviousStatus() {
        return previousStatus;
    }

    public void setPreviousStatus(ApplicationStatus previousStatus) {
        this.previousStatus = previousStatus;
    }

    public ApplicationStatus getNewStatus() {
        return newStatus;
    }

    public void setNewStatus(ApplicationStatus newStatus) {
        this.newStatus = newStatus;
    }

    public LocalDateTime getChangedAt() {
        return changedAt;
    }

    public void setChangedAt(LocalDateTime changedAt) {
        this.changedAt = changedAt;
    }

    public String getTrigger() {
        return trigger;
    }

    public void setTrigger(String trigger) {
        this.trigger = trigger;
    }

    public String getNotes() {
        return notes;
    }

    public void setNotes(String notes) {
        this.notes = notes;
    }
}
